# <center>Data Understanding</center>

In [23]:
# Importing the required Libraries
import pandas as pd

In [24]:
# Loading the dataset
df = pd.read_csv("data/predictive_maintenance.csv")

In [25]:
# Checking the shape of the dataset
print(f"The Shape of the dataset is {df.shape}")

The Shape of the dataset is (10000, 9)


In [26]:
# Checking the first five rows
df.head()

,UDI,Product ID,Type,Air temperature [K],Process temperature [K],Rotational speed [rpm],Torque [Nm],Tool wear [min],Target
0,1,M14860,M,298.1,308.6,1551,42.8,0,0
1,2,L47181,L,298.2,308.7,1408,46.3,3,0
2,3,L47182,L,298.1,308.5,1498,49.4,5,0
3,4,L47183,L,298.2,308.6,1433,39.5,7,0
4,5,L47184,L,298.2,308.7,1408,40.0,9,0


In [27]:
# Checking the column names
df.columns

Index(['UDI', 'Product ID', 'Type', 'Air temperature [K]',
       'Process temperature [K]', 'Rotational speed [rpm]', 'Torque [Nm]',
       'Tool wear [min]', 'Target'],
      dtype='object')

In [28]:
# Cleaning the column names
df.columns = (
    df.columns
    .str.strip()                          # remove leading and trailing spaces
    .str.lower()                          # convert to lowercase
    .str.replace(' ', '_')                # replace spaces with _
    .str.replace('[^a-zA-Z0-9_]', '', regex=True)  # remove special characters
)

# Checking the updated column names
print(df.columns)

Index(['udi', 'product_id', 'type', 'air_temperature_k',
       'process_temperature_k', 'rotational_speed_rpm', 'torque_nm',
       'tool_wear_min', 'target'],
      dtype='object')


In [29]:
# Checking the summary of the dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 9 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   udi                    10000 non-null  int64  
 1   product_id             10000 non-null  object 
 2   type                   10000 non-null  object 
 3   air_temperature_k      10000 non-null  float64
 4   process_temperature_k  10000 non-null  float64
 5   rotational_speed_rpm   10000 non-null  int64  
 6   torque_nm              10000 non-null  float64
 7   tool_wear_min          10000 non-null  int64  
 8   target                 10000 non-null  int64  
dtypes: float64(3), int64(4), object(2)
memory usage: 703.2+ KB


### Insights

- The dataset consists of 4 integer features, 3 float features, 2 categorical (object) features.

This indicates a mix of numerical and categorical data.

- Most of the features are numerical (temperature, speed, torque, tool wear), which are directly usable for machine learning models after scaling.

- The features type and product_id are categorical and will require encoding before being used in machine learning models.

- The target variable target is of integer type and represents a binary classification problem (0: No Failure, 1: Failure).

- All columns have 10,000 non-null values, indicating that there are no missing values in the dataset.

- The dataset appears to be clean and well-structured, requiring minimal preprocessing.

In [30]:
# Checking the statistical summary of numerical columns
df.describe()

,udi,air_temperature_k,process_temperature_k,rotational_speed_rpm,torque_nm,tool_wear_min,target
count,10000.00000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000
mean,5000.50000,300.004930,310.005560,1538.776100,39.986910,107.951000,0.033900
std,2886.89568,2.000259,1.483734,179.284096,9.968934,63.654147,0.180981
min,1.00000,295.300000,305.700000,1168.000000,3.800000,0.000000,0.000000
25%,2500.75000,298.300000,308.800000,1423.000000,33.200000,53.000000,0.000000
50%,5000.50000,300.100000,310.100000,1503.000000,40.100000,108.000000,0.000000
75%,7500.25000,301.500000,311.100000,1612.000000,46.800000,162.000000,0.000000
max,10000.00000,304.500000,313.800000,2886.000000,76.600000,253.000000,1.000000


### Insights
- The dataset contains 10,000 records with no missing values. The target variable is highly imbalanced, with only ~3.39% failure cases.
- Temperature features show low variance, indicating stable operating conditions, while rotational speed, torque, and tool wear exhibit significant variation and are likely important predictors.
- The UDI column appears to be an identifier and does not contribute to prediction. We will drop it

In [31]:
# Checking the statistical Summary of Categorical Columns
df.describe(exclude='number')

,product_id,type
count,10000,10000
unique,10000,3
top,L57163,L
freq,1,6000


### Insights
- The product_id column contains unique values for each record and does not contribute to prediction, hence it will be dropped.
- The type column is a categorical feature with three categories, with an imbalanced distribution where type L dominates with 60% of values. This feature will be retained and encoded during preprocessing.

In [32]:
# Checking for null values in the columns
df.isnull().sum()

udi                      0
product_id               0
type                     0
air_temperature_k        0
process_temperature_k    0
rotational_speed_rpm     0
torque_nm                0
tool_wear_min            0
target                   0
dtype: int64

### Insights
- As we seen earlier also using df.info() No null value exists. Therefore, no imputation is required

In [33]:
# Checking for duplicates
df.duplicated().sum()

np.int64(0)

### Insights

- No Duplicate rows in the dataset

In [34]:
# Saving the new dataset
df.to_csv("data/cleaned_data.csv", index=False)